### ISSUE TITLE
Errors for discriminated unions

### ISSUE THREAD

**Issue by zeeeeby** [LegendApp/legend-state#251]
For this example
```
type Data = {}
type State =
	| { state: "init" }
	| { state: "loading"; userId: number }
	| { state: "loaded"; userId: number; userData: Data  };

const state$ = observable<State>({ state: "init" });
state$.set({ state: "loading", userId: 4 });
```
I get this ts error
```
Argument of type '{ state: "loading"; userId: number; }' is not assignable to parameter of type '({ state: "init"; } & ((prev: { state: "loading"; userId: number; }) => { state: "loading"; userId: number; }) & ((prev: { state: "loaded"; userId: number; userData: Data; }) => { state: "loaded"; userId: number; userData: Data; })) | ... 20 more ... | undefined'.
  Type '{ state: "loading"; userId: number; }' is not assignable to type '(((prev: { state: "init"; }) => { state: "init"; }) & { state: "loading"; userId: number; } & ((prev: { state: "loaded"; userId: number; userData: Data; }) => { state: "loaded"; userId: number; userData: Data; })) | (((prev: { ...; }) => { ...; }) & ... 1 more ... & Promise<...>) | (Promise<...> & ... 1 more ... & (...'.
    Type '{ state: "loading"; userId: number; }' is not assignable to type 'Promise<{ state: "init"; }> & { state: "loading"; userId: number; } & Promise<{ state: "loaded"; userId: number; userData: Data; }>'.
      Type '{ state: "loading"; userId: number; }' is missing the following properties from type 'Promise<{ state: "init"; }>': then, catch, finally, [Symbol.toStringTag]ts(2345)
```

**Comment by jmeistrich** [comment]
Thanks @zeeeeby. We've rewritten the types from scratch for our upcoming version 3.0, and although it fixed this error we found another issue with discriminated unions using your example. That is now fixed and this example works correctly in version 2.2.0-next.58 if you want to try it. 

If you see any more type issues with that version please let me know! We are working on stabilizing it now towards the 3.0 release, which hopefully will be within a few weeks.

The new version is running well in production but it's not documented yet and we may want to rename some of the new APIs so you likely don't want to upgrade just yet... But it's coming soon :)

**Comment by jmeistrich** [comment]
I believe this is fixed in 3.0 alpha that I just released. But please let me know if it’s still not right.
